programs/manifest/src/state/market.rs

```diff
@@ -0,0 +1,1077 @@
+use bytemuck::{Pod, Zeroable};
+use hypertree::{
+    get_helper, get_mut_helper, trace, DataIndex, FreeList, PodBool, RBNode, RedBlackTree,
+    RedBlackTreeReadOnly, TreeReadOperations, NIL,
+};
+use solana_program::{entrypoint::ProgramResult, program_error::ProgramError, pubkey::Pubkey};
+use static_assertions::const_assert_eq;
+use std::mem::size_of;
+
+use crate::{
+    logs::{emit_stack, FillLog},
+    program::{assert_with_msg, ManifestError},
+    quantities::{BaseAtoms, GlobalAtoms, QuoteAtoms, QuoteAtomsPerBaseAtom, WrapperU64},
+    state::{
+        utils::{
+            assert_can_take, move_global_tokens_and_modify_resting_order, try_to_remove_from_global,
+        },
+        OrderType,
+    },
+    validation::{
+        get_vault_address, loaders::GlobalTradeAccounts, ManifestAccount, MintAccountInfo,
+    },
+};
+
+use super::{
+    claimed_seat::ClaimedSeat,
+    constants::{BLOCK_SIZE, MARKET_FIXED_SIZE},
+    order_type_can_rest,
+    utils::{assert_not_already_expired, get_now_slot, try_to_add_to_global},
+    DerefOrBorrow, DerefOrBorrowMut, DynamicAccount, RestingOrder, FREE_LIST_BLOCK_SIZE,
+    MARKET_FIXED_DISCRIMINANT,
+};
+
+pub struct AddOrderToMarketArgs<'a, 'info> {
+    pub market: Pubkey,
+    pub trader_index: DataIndex,
+    pub num_base_atoms: BaseAtoms,
+    pub price: QuoteAtomsPerBaseAtom,
+    pub is_bid: bool,
+    pub last_valid_slot: u32,
+    pub order_type: OrderType,
+    pub global_trade_accounts_opts: &'a [Option<GlobalTradeAccounts<'a, 'info>>; 2],
+}
+
+pub struct AddOrderToMarketResult {
+    pub order_sequence_number: u64,
+    pub order_index: DataIndex,
+    pub base_atoms_traded: BaseAtoms,
+    pub quote_atoms_traded: QuoteAtoms,
+}
+
+#[repr(C, packed)]
+#[derive(Default, Copy, Clone, Pod, Zeroable)]
+struct MarketUnusedFreeListPadding {
+    _padding: [u64; 9],
+    _padding2: [u8; 4],
+}
+// 4 bytes are for the free list, rest is payload.
+const_assert_eq!(
+    size_of::<MarketUnusedFreeListPadding>(),
+    FREE_LIST_BLOCK_SIZE
+);
+// Does not need to align to word boundaries because does not deserialize.
+
+#[repr(C)]
+#[derive(Default, Copy, Clone, Zeroable, Pod)]
+pub struct MarketFixed {
+    /// Discriminant for identifying this type of account.
+    pub discriminant: u64,
+
+    /// Version
+    version: u8,
+    base_mint_decimals: u8,
+    quote_mint_decimals: u8,
+    base_vault_bump: u8,
+    quote_vault_bump: u8,
+    _padding1: [u8; 3],
+
+    /// Base mint
+    base_mint: Pubkey,
+    /// Quote mint
+    quote_mint: Pubkey,
+
+    /// Base vault
+    base_vault: Pubkey,
+    /// Quote vault
+    quote_vault: Pubkey,
+
+    /// The sequence number of the next order.
+    order_sequence_number: u64,
+
+    /// Num bytes allocated as RestingOrder or ClaimedSeat or FreeList. Does not
+    /// include the fixed bytes.
+    num_bytes_allocated: u32,
+
+    /// Red-black tree root representing the bids in the order book.
```

Comment:
There was talk of renaming red-black trees? Assuming a todo

Reply:
Just the name toolbox was the problem, we can leave the name here or not, I dont care